pkg/cmd/server/origin/handlers.go

```diff
@@ -291,3 +297,112 @@ func assetServerRedirect(handler http.Handler, assetPublicURL string) http.Handl
 		handler.ServeHTTP(w, req)
 	})
 }
+
+func (c *MasterConfig) impersonationFilter(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestedSubject := req.Header.Get("Impersonate-User")
+		if len(requestedSubject) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		resource, namespace, name, err := parseRequestedSubject(requestedSubject)
+		if err != nil {
+			forbidden(err.Error(), nil, w, req)
+			return
+		}
+
+		ctx, exists := c.RequestContextMapper.Get(req)
+		if !exists {
+			forbidden("context not found", nil, w, req)
+			return
+		}
+
+		actingAsAttributes := &authorizer.DefaultAuthorizationAttributes{
+			Verb:         "impersonate",
+			APIGroup:     resource.Group,
+			Resource:     resource.Resource,
+			ResourceName: name,
+		}
+		authCheckCtx := kapi.WithNamespace(ctx, namespace)
+
+		allowed, reason, err := c.Authorizer.Authorize(authCheckCtx, actingAsAttributes)
+		if err != nil {
+			forbidden(err.Error(), actingAsAttributes, w, req)
+			return
+		}
+		if !allowed {
+			forbidden(reason, actingAsAttributes, w, req)
+			return
+		}
+
+		switch resource {
```

Comment:
> Is there any point to having this be a common function (or more closely shared). The groups lookups look somewhat fragile.

Maybe?  We apply it during authorization as a filter, so it might be possible to re-swizzle it.  Tech debt issue for someone wanting to learn about the authn/authz layers?